configcatclient/rolloutevaluator.py

```diff
@@ -48,174 +169,553 @@ def evaluate(self, key, user, default_value, default_variation_id, settings):  #
+        user = context.user
+        key = context.key
+
+        segment_index = segment_condition.get(SEGMENT_INDEX)
+        segment = segments[segment_index]
+        segment_name = segment.get(SEGMENT_NAME, '')
+        segment_comparator = segment_condition.get(SEGMENT_COMPARATOR)
+        segment_comparison_rules = segment.get(SEGMENT_RULES, [])
+
+        if user is None:
+            if not context.is_missing_user_object_logged:
+                self.log.warning('Cannot evaluate targeting rules and %% options for setting \'%s\' '
+                                 '(User Object is missing). '
+                                 'You should pass a User Object to the evaluation methods like `get_value()` '
+                                 'in order to make targeting work properly. '
+                                 'Read more: https://configcat.com/docs/advanced/user-object/',
+                                 key, event_id=3001)
+                context.is_missing_user_object_logged = True
+            if log_builder:
+                log_builder.append("User %s '%s' " % (self.SEGMENT_COMPARATOR_TEXTS[segment_comparator], segment_name))
+            return False, 'cannot evaluate, User Object is missing'
+
+        # IS IN SEGMENT, IS NOT IN SEGMENT
+        if segment_comparator in [SegmentComparator.IS_IN, SegmentComparator.IS_NOT_IN]:
+            if log_builder:
+                log_builder.append("User %s '%s'" %
+                                   (self.SEGMENT_COMPARATOR_TEXTS[segment_comparator], segment_name))
+                log_builder.new_line('(').increase_indent()
+                log_builder.new_line("Evaluating segment '%s':" % segment_name)
+
+            # Set initial condition result based on comparator
+            segment_condition_result = segment_comparator == SegmentComparator.IS_IN
+
+            # Evaluate segment rules (logically connected by AND)
+            first_segment_rule = True
+            error = None
+            for segment_comparison_rule in segment_comparison_rules:
+                if first_segment_rule:
+                    if log_builder:
+                        log_builder.new_line('- IF ')
+                        log_builder.increase_indent()
+                    first_segment_rule = False
+                else:
+                    if log_builder:
+                        log_builder.new_line('AND ')
+
+                result, error = self._evaluate_comparison_rule_condition(segment_comparison_rule, context,
+                                                                         segment_name, salt, log_builder)
+                if log_builder:
+                    log_builder.append('=> {}'.format('true' if result else 'false'))
+                    if not result:
+                        log_builder.append(', skipping the remaining AND conditions')
+
+                if not result:
```

Comment:
I had to read this multiple times to understand why is this solution correct. :D
Elegant and simple! Nice!